### PR TITLE
Add async XLinq document/element loading and saving

### PR DESCRIPTION
**Pull request by sepidehMS**
This PR resolves https://github.com/dotnet/corefx/issues/2869. Merges https://github.com/dotnet/corefx/pull/2436 from future branch into master and exposes the new APIs for netcoreapp1.1.

@danmosemsft @stephentoub @weshaggard

### CODE REVIEW

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XCData.cs [hunk 1]
```diff
@@ -49,6 +52,27 @@ public override void WriteTo(XmlWriter writer)
...
+        /// <param name="cancellationToken">
+        /// The CancellationToken to use to request cancellation of this operation.
+        /// </param>
+        /// <returns>
+        /// A Task that represents the eventual completion of the operation.
+        /// </returns>
+        public override Task WriteToAsync(XmlWriter writer, CancellationToken cancellationToken)
+        {
+            if (writer == null)
+                throw new ArgumentNullException("writer");
```
Comment (justinvp) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XCData.cs]: Nit: `nameof(writer)` (I think we switched over to using `nameof` throughout after these changes were already checked-in to the future branch). Applies to a few other files in this PR.

File: src/System.Private.Xml.Linq/tests/misc/LoadSaveAsyncTests.cs [hunk 2]
```diff
@@ -0,0 +1,205 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (justinvp) [inline src/System.Private.Xml.Linq/tests/misc/LoadSaveAsyncTests.cs]: This should be replaced with the new license header

File: src/System.Xml.XDocument/ref/System.Xml.XDocument.cs [hunk 3]
```diff
@@ -121,6 +121,7 @@ public partial class XCData : System.Xml.Linq.XText
         public XCData(System.Xml.Linq.XCData other) : base(default(string)) { }
         public override System.Xml.XmlNodeType NodeType { get { throw null; } }
         public override void WriteTo(System.Xml.XmlWriter writer) { }
+        public override System.Threading.Tasks.Task WriteToAsync(System.Xml.XmlWriter writer, System.Threading.CancellationToken cancellationToken) { throw null; }
```
Comment (justinvp) [inline src/System.Xml.XDocument/ref/System.Xml.XDocument.cs]: I believe all of these new APIs in the contracts need to be surrounded by `#if netcoreapp11`. See some of the other PRs that contribute to #2869 like #12600 and #13828.

**Comment by sepidehMS** [comment]
@justinvp Thanks for your comments. Made according changes.

File: src/System.Xml.XDocument/ref/System.Xml.XDocument.cs [hunk 4]
```diff
@@ -178,6 +184,11 @@ public partial class XDocument : System.Xml.Linq.XContainer
         public static System.Xml.Linq.XDocument Load(string uri, System.Xml.Linq.LoadOptions options) { throw null; }
         public static System.Xml.Linq.XDocument Load(System.Xml.XmlReader reader) { throw null; }
         public static System.Xml.Linq.XDocument Load(System.Xml.XmlReader reader, System.Xml.Linq.LoadOptions options) { throw null; }
+#if netcoreapp11
+        public static System.Threading.Tasks.Task<System.Xml.Linq.XDocument> LoadAsync(System.IO.Stream stream, System.Xml.Linq.LoadOptions options, System.Threading.CancellationToken cancellationToken) { throw null; }
```
Comment (danmosemsft) [inline src/System.Xml.XDocument/ref/System.Xml.XDocument.cs]: do you know why we don't put 'async' keyword in the ref file?

Comment (JonHanna) [inline src/System.Xml.XDocument/ref/System.Xml.XDocument.cs]: That would create another private type for the state machine that isn't necessary.

Comment (danmosemsft) [inline src/System.Xml.XDocument/ref/System.Xml.XDocument.cs]: Makes sense.

**Comment by sepidehMS** [comment]
Any additional comments?

**Comment by stephentoub** [comment]
I'll take a look today...

**Comment by sepidehMS** [comment]
Thanks for reviewing.

**Pull request merged**